Include Turbolinks if it is installed

wnm opened the pull request:
Turbolinks is overwriting the redirect_to method of ActionController::Redirecting.

Devise only includes ActionController::Redirecting

So, if turbolinks is installed, it doesn't work will with devise.

This should fix it.

rafaelfranca commented:
The failure app should not know about turbolinks. I know it will redirect without turbolinks in the case the authentication fails, but devise should not know anything about turbolinks and fully rendering a failure page is not a performance problem to justify this coupling. Thank you for the pull request.

wnm commented:
@rafaelfranca it's not a performance issue, no. But using Turbolinks with the wrappers for Android and iOS requires Turbolinks redirects. The Webview doesn't follow normal redirects. 

But I understand you don't want to include this. 

For anyone stumbling opon this issue, put this in your devise config:

```
module Devise
  class FailureApp
    include Turbolinks::Controller
  end
end
```

rafaelfranca commented:
I see. We could take a different approach. Instead of adding that in the controller we could allow people to extend the devise failure app without monkey patching. Something like:

```ruby
ActiveSupport.run_load_hooks(:devise_failure_app)
```

in the controller and

```ruby
ActiveSupport.on_load(:devise_failure_app)
  include Turbolinks::Controller
end
```

tegon commented:
I agree we can provide a lazy load hook. And it seems it shouldn't be much hard to do it.
If I understood it right, we just need to add `ActiveSupport.run_load_hooks(:devise_failure_app, self)` to the end of the [failure app](https://github.com/plataformatec/devise/blob/master/lib/devise/failure_app.rb) class.

@wnm If you're willing to change this implementation to use a lazy load hook, we'd happily accept the pull request. Otherwise we can implement it in another one.

Some related resources:

http://api.rubyonrails.org/classes/ActiveSupport/LazyLoadHooks.html
https://simonecarletti.com/blog/2011/04/understanding-ruby-and-rails-lazy-load-hooks/

wnm commented:
@tegon I updated the pull request:

- Added `ActiveSupport.run_load_hooks(:devise_failure_app, self)` to the end of failure app class.
- Included config to load Turbolinks::Controller in devise initilizer template (but commented out)

Hope I understood everything correctly?

tegon commented on lib/generators/templates/devise.rb:
```diff
@@ -280,4 +280,11 @@
+  # ==> Turbolinks configuration
+  # If your app is using Turbolinks, Turbolinks::Controller needs to be included to make redirection work correctly:
+  #
+  # ActiveSupport.on_load(:devise_failure_app)
```
I believe we're missing a `do` block here
```diff
- # ActiveSupport.on_load(:devise_failure_app)
+ # ActiveSupport.on_load(:devise_failure_app) do

tegon commented:
@wnm Thanks! I think that's basically it. I just think we should have a test for this.
One way we can test it is to create a simple module with a method and include it in the failure app using this hook, then we can check if the failure app responds to that method.
If you need help with this test, let me know.

wnm commented:
@tegon not sure if the tests are in the right place. let me know if they are not...

tegon commented:
@wnm Looks good, thanks again!

dsandstrom commented:
This doesn't seem to be required for Rails 6 and may cause confusion. Was trying to troubleshoot and tried to toggle this config on, but it errors for me.  My issue ended up being a conflict with my existing user routes.

I'm not using the turbolinks gem so there's no `Controller` to include. Instead, I installed turbolinks with yarn and it gets loaded in application.js.

Error:
```
uninitialized constant Turbolinks
```

More info about rails js: https://prathamesh.tech/2019/08/26/understanding-webpacker-in-rails-6/

Just an fyi. Thanks for the PR/gems.
